draft-ietf-quic-transport.md

```diff
@@ -860,39 +860,93 @@ different IP or port at either endpoint, due to NAT rebinding or mobility, as
+packets to the largest minimum size across all of the QUIC versions they
+support.
+
+
 ## Version Negotiation {#version-negotiation}
 
 QUIC's connection establishment begins with version negotiation, since all
 communication between the endpoints, including packet and frame formats, relies
 on the two endpoints agreeing on a version.
 
-A QUIC connection begins with a client sending a handshake packet. The details
-of the handshake mechanisms are described in {{handshake}}, but all of the
-initial packets sent from the client to the server MUST use the long header
-format and MUST specify the version of the protocol being used.
-
-When the server receives a packet from a client with the long header format, it
-compares the client's version to the versions it supports.
-
-If the version selected by the client is not acceptable to the server, the
-server discards the incoming packet and responds with a Version Negotiation
-packet ({{packet-version}}).  This includes a list of versions that the server
-will accept.
+A QUIC connection begins with a client sending a client initial packet
+({{packet-client-initial}}). The details of the handshake mechanisms are
+described in {{handshake}}, but all of the initial packets sent from the client
+to the server MUST use the long header format - which includes the version of
+the protocol being used - and they MUST be padded to at least 1200 octets.
 
-A server sends a Version Negotiation packet for every packet that it receives
-with an unacceptable version.  This allows a server to process packets with
-unsupported versions without retaining state.  Though either the initial client
-packet or the version negotiation packet that is sent in response could be lost,
-the client will send new packets until it successfully receives a response.
+The server receives this packet and determines that it potentially creates a new
+connection.  It then checks whether it understands the version that the client
+has selected.
 
 If the packet contains a version that is acceptable to the server, the server
 proceeds with the handshake ({{handshake}}).  This commits the server to the
 version that the client selected.
 
-When the client receives a Version Negotiation packet from the server, it should
-select an acceptable protocol version.  If the server lists an acceptable
-version, the client selects that version and reattempts to create a connection
-using that version.  Though the contents of a packet might not change in
+
+### Sending and Handling Version Negotiation Packets {#send-vn}
+
+If the version selected by the client is not acceptable to the server, the
+server responds with a Version Negotiation packet ({{packet-version}}).  This
+includes a list of versions that the server will accept.
+
+A server sends a Version Negotiation packet for any packet with an unaccepable
+version if that packet could create a new connection.  This allows a server to
```

Comment:
Does it make sense to specify only sending VN in response to client initial or are there other types you have in mind for "that packet could create a new connection."?

Reply:
I think Ian's point is that it's not clear what "might_create_a_connection" means. I am not clear about this either, TBH. How does a server know that a received packet might create a connection if the version is unspoken? I think all you need is if (!acceptable(version)) { send_vn() }

Reply:
No short packet can start a connection.  Long packets with connection IDs that are currently active inherently can't start a different connection.  So, sticking with the pseudo-code, `might_create_a_connection := !ActiveConnectionIDs.Contains(packet.connection_id) && (packet.header.IsLongFormat)`

Reply:
If there's an active connection with this ID, you will match the connection earlier, attempt to decrypt the packet with the stored key and fail, and drop the packet on the floor. Even if you check the version before you try to decrypt you'd send a VN packet in response.

So this boils down to `if ((flags & 0x80) && !acceptable(version)) { send_vn() }`, since whether there's an active connection or not, we still send a VN packet. In which case, I think the long header check is unnecessary.

Simply checking the version seems adequate, no?

Reply:
The text (that this should have referred to) boils down to almost exactly what you describe, but with a little more nuance.  "Might have created a connection" is defined further up (in this self-same PR), but I neglected to cross-reference it properly.

In short, the condition is expanded to include one more check: that the packet is big enough to generate a new connection on any version that the server supports.  This means that if you are doing 0-RTT, then you might get a Version Negotiation packet only in response to the Client Initial and any 0-RTT packet that is large enough.  This avoids the server sending Version Negotiation in response to smaller packets (and thereby acting as an amplifier).